apps/omg_eth/lib/omg_eth/ethereum_height_monitor.ex

```diff
@@ -0,0 +1,237 @@
+          tref: reference() | nil,
+          alarm_module: module(),
+          event_bus: module(),
+          ethereum_height: integer(),
+          synced_at: DateTime.t(),
+          connection_alarm_raised: boolean(),
+          stall_alarm_raised: boolean(),
+          events: events_t()
+        }
+
+  defstruct check_interval_ms: 10_000,
+            stall_threshold_ms: 20_000,
+            tref: nil,
+            alarm_module: nil,
+            event_bus: nil,
+            ethereum_height: 0,
+            synced_at: nil,
+            connection_alarm_raised: false,
+            stall_alarm_raised: false,
+            events: []
+
+  #
+  # GenServer APIs
+  #
+
+  def start_link(args) do
+    GenServer.start_link(__MODULE__, args, name: __MODULE__)
+  end
+
+  @doc """
+  Return current events that are associated with `EthereumHeightMonitor`.
+  """
+  @spec get_events() :: {:ok, events_t()}
+  def get_events() do
+    GenServer.call(__MODULE__, :get_events)
+  end
+
+  #
+  # GenServer behaviors
+  #
+
+  def init(opts) do
+    _ = Logger.info("Starting Ethereum height monitor.")
+    _ = install_alarm_handler()
+
+    state = %__MODULE__{
+      check_interval_ms: Application.fetch_env!(:omg_eth, :ethereum_height_check_interval_ms),
+      stall_threshold_ms: Application.fetch_env!(:omg_eth, :ethereum_stalled_sync_threshold_ms),
+      synced_at: DateTime.utc_now(),
+      alarm_module: Keyword.fetch!(opts, :alarm_module),
+      event_bus: Keyword.fetch!(opts, :event_bus)
+    }
+
+    {:ok, tref} = :timer.send_after(state.check_interval_ms, :check_new_height)
+    {:ok, %{state | tref: tref}}
+  end
+
+  def handle_info(:check_new_height, state) do
+    height = fetch_height()
+    stalled? = stalled?(height, state.ethereum_height, state.synced_at, state.stall_threshold_ms)
+
+    :ok = broadcast_on_new_height(state.event_bus, state.ethereum_height, height)
+    _ = connection_alarm(state.alarm_module, state.connection_alarm_raised, height)
+    _ = stall_alarm(state.alarm_module, state.stall_alarm_raised, stalled?)
+
+    state =
+      case height > state.ethereum_height do
+        true -> %{state | ethereum_height: height, synced_at: DateTime.utc_now()}
+        false -> state
+      end
+
+    {:ok, tref} = :timer.send_after(state.check_interval_ms, :check_new_height)
+    {:noreply, %{state | tref: tref}}
+  end
+
+  def handle_call(:get_events, _from, state) do
+    {:reply, {:ok, state.events}, state}
+  end
+
+  #
+  # Handle incoming alarms
+  #
+  # These functions are called by the AlarmHandler so that this monitor process can update
+  # its internal state according to the raised alarms. Mainly these handlers do 2 things:
+  #
+  # 1. Reflect the internal state so that it does not re-raise an existing alarm
+  # 2. Maintain the list of events that the alarm corresponds to
+  #
+  def handle_cast({:set_alarm, :ethereum_connection_error}, state) do
+    events = [%Event.EthereumConnectionError{} | state.events]
```

Comment:
how are you using the events list?

Reply:
`/status.get` is calling `EthereumHeightMonitor.get_events()`

Another approach I thought but didn't pick is to render the events when `EthereumHeightMonitor.get_events()` is called, but that means each call to `/status.get` will build a new `%Event.EthereumConnectionError{}` and/or `%Event.EthereumStalledSync{}` which I felt a bit too unnecessary.

Or lemme know if you have other suggestions!